### ISSUE TITLE
Tests: Refactor lib/plugins/package/lib/packageService.test.js 

### ISSUE THREAD

**Issue by medikoo** [serverless/serverless#8594]
_Most of the tests that currently cover core functionalities depend heavily on (and sometimes test) internal implementation characteristics. While they should be testing whether given implementation produces desired outcome (treating its implementation more as  a black box)._

_Current state of things is problematic for eventual internal improvements and refactors which occasionally we want to introduce, as in most of such cases proposed improvements need to be accompanied with counterproductive numerous updates to tests which are covering otherwise not altered functionalities ._

_At some point we've introduced a new (black box based) way of testing the internals. It's through `runServerless` utlity, which allows us create natural (as in real world) `serverless` instance, and inspect the produced outcome for chosen command. More details here: https://github.com/serverless/serverless/tree/master/test#unit-tests_

--- 

This is issue is about refactoring [test/unit/lib/plugins/package/lib/packageService.test.js](https://github.com/serverless/serverless/blob/44d398694fc803bb45a07e7bcc0d6cd9e572a704/test/unit/lib/plugins/package/lib/packageService.test.js) to `runServerless` based variant.

To make refactor relatively easy:
- Spec for new tests is fully defined and is already committed in. It's in scope of this _describe_ block: https://github.com/serverless/serverless/blob/44d398694fc803bb45a07e7bcc0d6cd9e572a704/test/unit/lib/plugins/package/lib/packageService.test.js#L456-L659
- Each test slot indicates which test it is about to replace
- It is indicated where (and how many) exactly `runServerless` runs are expected to happen, against which fixture and command they should be based

For every refactored test, the old test should be removed

Refactor can be submitted with one PR, but partial refactors that cover some of tests are also very welcome!

**Comment by scadu** [comment]
If there are no volunteers, I'd pick up this one.

**Comment by medikoo** [comment]
@scadu that'll be great. Thank you!

**Comment by scadu** [comment]
@medikoo Hi,

A small update, since this issue is blocking #8581 and I feel stuck. My brain is a mush this week.
I started tinkering and the current, erm, progress can be seen [here](https://github.com/scadu/serverless/pull/2) (a draft PR against my own fork). Nothing useful so far.

It seems like Go-related tests weren't even used due to the typo (`process.platfrom` instead of `process.platform`).
Anyway, the first part (`service wide`) is barely finished. I couldn't get `exclude` working according to the requirements - seems to be conflict between service and function scope where the same path is excluded and included. Most probably I miss something obvious (see; https://github.com/scadu/serverless/blob/b4f88b13e4b4e73cfde86ca77ee2ba1f917e88c0/test/unit/lib/plugins/package/lib/packageService.test.js#L489-L491). 
Also, I have some doubts regarding Go-related tests:
1. Make sure that handler file is marked executable on Windows: https://github.com/scadu/serverless/blob/b4f88b13e4b4e73cfde86ca77ee2ba1f917e88c0/test/unit/lib/plugins/package/lib/packageService.test.js#L489-L491. For now I just pointed to the `main.go`, but I've tested it locally against binary (`bin/main`). Not sure if building the Go binary is something we'd like to do. 
2. Make sure normalized handler path is used on Windows: 
https://github.com/scadu/serverless/blob/b4f88b13e4b4e73cfde86ca77ee2ba1f917e88c0/test/unit/lib/plugins/package/lib/packageService.test.js#L515-L521. I haven't figured it out how to validate it properly against existing fixtures.

I just barely touched the second stage (`individually`), but I ended up with the following just by removing the `.skip`.
I'd need to check the `PluginManager` for some clues, since it also might be something trivial.
```
  1) lib/plugins/package/lib/packageService.test.js
       TODO: individually
         "before all" hook for "should exclude custom plugins localPath":
     TypeError: Plugin is not a constructor
```

tldr: not much progress so far. I'll try to dig deeper tomorrow or the day after.

**Comment by medikoo** [comment]
@scadu thanks for asking

> I couldn't get exclude working according to the requirements - seems to be conflict between service and function scope where the same path is excluded and included. 

Indeed the proposed patterns were a bit confusing. I've improved that I believe with this commit: https://github.com/serverless/serverless/commit/4505deebf2fbc14eaa90c0f669d68be84394292c Let me know if after taking this in it works without issues

> Make sure that handler file is marked executable on Windows

This test should be run only windows (as it's only in this env where it's guaranteed to pass), and we should (1) _unpack_ generated zip artifact (2) get go handler stats (simply via `fs.promises.stat`) (3) confirm that `stats.mode` have execution bit

(at least I think that `listZipFiles` util won't give a mean to inspect file mode (?) Maybe it can be extended to provide additional meta if requested (?))

> Make sure normalized handler path is used on Windows

Old test as I checked simply confirmed that normalized path to handler was passed to zip function. I think we can skip that test totally, as we normalize all paths as they come. I removed this test slot in master 

> I just barely touched the second stage (individually), but I ended up with the following just by removing the .skip

It was due to custom _plugins_ as setup in fixture not exporting functions. I've fixed that with https://github.com/serverless/serverless/commit/402cfced70f404e1bd28d244d86e81ee5464099a

Great thanks for all the questions, and let me know if there's anything else

**Comment by scadu** [comment]
> @scadu thanks for asking
> 
> > I couldn't get exclude working according to the requirements - seems to be conflict between service and function scope where the same path is excluded and included.
> 
> Indeed the proposed patterns were a bit confusing. I've improved that I believe with this commit: [4505dee](https://github.com/serverless/serverless/commit/4505deebf2fbc14eaa90c0f669d68be84394292c) Let me know if after taking this in it works without issues

It looks like the exclude `dir1/subdir2/subsubdir3` on the function level is ignored and the include definition from the service level is taken (`include: ['dir1/subdir2/**', '!dir1/subdir2/subsubdir1']`) - the package includes `dir1/subdir2/subsubdir2` and `dir1/subdir2/subsubdir3`. So at least now I see the rationale behind patterns (#8581). Still, the inheritance seems to be tricky here.

> 
> > Make sure that handler file is marked executable on Windows
> 
> This test should be run only windows (as it's only in this env where it's guaranteed to pass), and we should (1) _unpack_ generated zip artifact (2) get go handler stats (simply via `fs.promises.stat`) (3) confirm that `stats.mode` have execution bit
> 
> (at least I think that `listZipFiles` util won't give a mean to inspect file mode (?) Maybe it can be extended to provide additional meta if requested (?))

I did something like this in https://github.com/scadu/serverless/blob/b4f88b13e4b4e73cfde86ca77ee2ba1f917e88c0/test/unit/lib/plugins/package/lib/packageService.test.js#L507. Not sure if that would be okay.

> 
> > Make sure normalized handler path is used on Windows
> 
> Old test as I checked simply confirmed that normalized path to handler was passed to zip function. I think we can skip that test totally, as we normalize all paths as they come. I removed this test slot in master
> 
> > I just barely touched the second stage (individually), but I ended up with the following just by removing the .skip
> 
> It was due to custom _plugins_ as setup in fixture not exporting functions. I've fixed that with [402cfce](https://github.com/serverless/serverless/commit/402cfced70f404e1bd28d244d86e81ee5464099a)
> 
> Great thanks for all the questions, and let me know if there's anything else

**Comment by scadu** [comment]
> > @scadu thanks for asking
> > > I couldn't get exclude working according to the requirements - seems to be conflict between service and function scope where the same path is excluded and included.
> > 
> > 
> > Indeed the proposed patterns were a bit confusing. I've improved that I believe with this commit: [4505dee](https://github.com/serverless/serverless/commit/4505deebf2fbc14eaa90c0f669d68be84394292c) Let me know if after taking this in it works without issues
> 
It looks like the exclude `dir1/subdir2/subsubdir3` on the function level is ignored and the include definition from the service level is taken (`include: ['dir1/subdir2/**', '!dir1/subdir2/subsubdir1']`) - the package includes `dir1/subdir2/subsubdir2` and `dir1/subdir2/subsubdir3`. So at least now I see the rationale behind patterns (#8581). Still, the inheritance seems to be tricky here.
> 
> > > Make sure that handler file is marked executable on Windows
> > 
> > 
> > This test should be run only windows (as it's only in this env where it's guaranteed to pass), and we should (1) _unpack_ generated zip artifact (2) get go handler stats (simply via `fs.promises.stat`) (3) confirm that `stats.mode` have execution bit
> > (at least I think that `listZipFiles` util won't give a mean to inspect file mode (?) Maybe it can be extended to provide additional meta if requested (?))
> 
I did something like this in https://github.com/scadu/serverless/blob/b4f88b13e4b4e73cfde86ca77ee2ba1f917e88c0/test/unit/lib/plugins/package/lib/packageService.test.js#L507. Not sure if that would be okay.
> 
> > > Make sure normalized handler path is used on Windows
> > 
> > 
> > Old test as I checked simply confirmed that normalized path to handler was passed to zip function. I think we can skip that test totally, as we normalize all paths as they come. I removed this test slot in master

👍 
> > > I just barely touched the second stage (individually), but I ended up with the following just by removing the .skip
> > 
> > 
> > It was due to custom _plugins_ as setup in fixture not exporting functions. I've fixed that with [402cfce](https://github.com/serverless/serverless/commit/402cfced70f404e1bd28d244d86e81ee5464099a)
> > Great thanks for all the questions, and let me know if there's anything else

Seems to be working if AWS credentials are set.

**Comment by medikoo** [comment]
> It looks like the exclude dir1/subdir2/subsubdir3 on the function level is ignored and the include definition from the service level is taken

Indeed, case with `exclude` and `include` rules it that first _exclude_ rules are resolved and then _include_ rules. So in this case `'dir1/subdir2/**'` _inclusion_ is shadowed `'dir1/subdir2/subsubdir3'` _exclusion.

I've updated the fixture and comments, and hopefully now everything would work as expected -> https://github.com/serverless/serverless/commit/77fbc73724c632d2388ae4b1d53a2649b0d520f2

> I did something like this in https://github.com/scadu/serverless/blob/b4f88b13e4b4e73cfde86ca77ee2ba1f917e88c0/test/unit/lib/plugins/package/lib/packageService.test.js#L507. Not sure if that would be okay.

This doesn't seem to validate what we're after, note that `expect` takes one argument. so you're just confirming that  `fnIndividualZippedFiles` includes 'main.go`.  While what we need to confirm here, is that not only it's packaged, but that it's packaged with expected mode.

> Seems to be working if AWS credentials are set.

It needs stubbing of AWS requests, as e.g. here: https://github.com/serverless/serverless/blob/77fbc73724c632d2388ae4b1d53a2649b0d520f2/test/unit/lib/plugins/package/lib/packageService.test.js#L421

**Comment by scadu** [comment]
@medikoo yep, after the latest changes (77fbc73) it started working.

I'm looking into file permissions now. JSZip sets `unixPermissions` as null when unpacking on Windows, so not much luck here. I'll check if we could inspect file attributes and make sure they were left untouched.

**Comment by scadu** [comment]
I think it'd be covered in https://github.com/scadu/serverless/pull/2/commits/c8068394e03f4c26fd9f6c8f3b85ba6aaaeb0900.
Now I'm trying to do the following:
* `individually`
  * should ignore `package.artifact` if `package.individually`
  * should support `package.individually`
* `pre-prepared artifact`
  *  should support `package.artifact`
  * should ignore `package.artifact` if `functions[].package.individually`
  * should support `functions[].package.artifact`
  
I suspect that `runServerless` should be called separately for each of them and then compare with a generated configuration. Am I right?

**Comment by medikoo** [comment]
> I suspect that runServerless should be called separately for each of them and then compare with a generated configuration. Is that right @medikoo?

@scadu good question. I've preconfigured all `runServerless` runs which I believe are needed. Still again I might have missed something. If you feel that some thinigs cannot be tested well in context of preconfigured runs, please explain, and we'll find the right solution

All cases in `pre-prepared artifact` can be tested with one run (in config we can have following properties set: `package.artifcact`, `functions.a.package.individually`, and `functions.b.package[individually & artifact]`)

**Comment by scadu** [comment]
So far I've opened PR that doesn't cover all tests (I haven't touched the codebase since the latest comment and I still have no clear idea how to do it), but I wouldn't like to block #8581.
